panther_hardware_interfaces/src/panther_imu.cpp

```diff
@@ -0,0 +1,513 @@
+  params_.remove_gravity_vector = hardware_interface::parse_bool(remove_gravity_vector);
+}
+
+void PantherImuSensor::CheckMadgwickFilterWorldFrameParam()
+{
+  const auto world_frame = info_.hardware_parameters.at("world_frame");
+
+  auto it = kImuWorldFramesMap.find(world_frame);
+  if (it == kImuWorldFramesMap.end()) {
+    throw std::runtime_error(
+      "The parameter world_frame was set to invalid value. "
+      "Valid values are 'enu', 'ned' and 'nwu'. Setting to 'enu'.");
+  }
+  world_frame_ = it->second;
+}
+
+void PantherImuSensor::SetInitialValues()
+{
+  imu_sensor_state_.resize(
+    info_.sensors[0].state_interfaces.size(), std::numeric_limits<double>::quiet_NaN());
+}
+
+void PantherImuSensor::Calibrate()
+{
+  if (spatial_ == nullptr) {
+    throw std::runtime_error("Imu hardware is not active yet!");
+  }
+
+  spatial_->zero();
+  // The API call returns directly, so we "enforce" the recommended 2 sec
+  // here. See: https://github.com/ros-drivers/phidgets_drivers/issues/40
+
+  // FIXME: Ideally we'd use an rclcpp method that honors use_sim_time here,
+  // but that doesn't actually exist.  Once
+  // https://github.com/ros2/rclcpp/issues/465 is solved, we can revisit this.
+  std::this_thread::sleep_for(std::chrono::seconds(2));
+}
+
+bool PantherImuSensor::IsParamDefined(const std::string & param_name)
+{
+  return info_.hardware_parameters.find(param_name) != info_.hardware_parameters.end();
+}
+
+void PantherImuSensor::ConfigureCompassParams()
+{
+  if (
+    IsParamDefined("cc_mag_field") and IsParamDefined("cc_offset0") and
+    IsParamDefined("cc_offset1") and IsParamDefined("cc_offset2") and IsParamDefined("cc_gain0") and
+    IsParamDefined("cc_gain1") and IsParamDefined("cc_gain2") and IsParamDefined("cc_t0") and
+    IsParamDefined("cc_t1") and IsParamDefined("cc_t2") and IsParamDefined("cc_t3") and
+    IsParamDefined("cc_t4") and IsParamDefined("cc_t5")) {
+    ReadCompassParams();
+
+    spatial_->setCompassCorrectionParameters(
+      params_.cc_mag_field, params_.cc_offset0, params_.cc_offset1, params_.cc_offset2,
+      params_.cc_gain0, params_.cc_gain1, params_.cc_gain2, params_.cc_t0, params_.cc_t1,
+      params_.cc_t2, params_.cc_t3, params_.cc_t4, params_.cc_t5);
+  } else {
+    RCLCPP_INFO(logger_, "No compass correction params found. Skipping...");
+  }
+}
+
+void PantherImuSensor::ConfigureHeating()
+{
+  if (IsParamDefined("heating_enabled")) {
+    params_.heating_enabled =
+      hardware_interface::stod(info_.hardware_parameters["heating_enabled"]);
+    spatial_->setHeatingEnabled(params_.heating_enabled);
+  } else {
+    RCLCPP_INFO(logger_, "No heating enabled found. Skipping...");
+  }
+}
+
+void PantherImuSensor::ConfigureMadgwickFilter()
+{
+  ReadMadgwickFilterParams();
+
+  filter_.setWorldFrame(world_frame_);
+  filter_.setAlgorithmGain(params_.gain);
+  filter_.setDriftBiasGain(params_.zeta);
+}
+
+void PantherImuSensor::SpatialDataCallback(
```

Comment:
Could we refactor this method to be shorter by introducing additional smaller methods?